### DISCUSSION TITLE
Multiple network interfaces

### DISCUSSION THREAD

**Discussion by lazna** [dannagle/PacketSender#290] [General]
Have multiple network interface on my windows 10 machine . When send broadcast packet from command line

`packetsender.com -quw 2000 255.255.255.255 10001 "01 00 00 00"`

it was not going to interface I need. To what interface is packet sent if not -B option specified? Is there any way the packet will be send to ALL present network interfaces?

**Comment by dannagle** [comment]
Hello,

If the socket is not bound to a particular IP address, the chosen network card for sending out 255.255.255.255 broadcasts is random. It might go out whatever card has a gateway attached, but that is not reliable.

You have 2 ways to go about fixing this...
 - Bind to an address of the desired network card before sending out the broadcast
   - Note that a card can have multiple IPs. Choose any from the card.
 - Disable all other network cards before sending. This will force the socket to use the desired card as it will be the only choice. 

If you want a broadcast to go out all your network cards, the only reliable way to do this is to systematically create sockets that bind to an address from each network card and then send an individual broadcast. These network interfaces could be completely isolated networks and need their own broadcast packet (e.g. 1 socket for WiFi, another socket for Ethernet, etc).

I hope that helps,
Dan

**Reply by lazna** [reply]
Thanks for reply. Changing default iface from 'random' to 'all' could make packetsender scripting much easier. Now I have to list all ifaces, than list its IP addresses, choose one on each iface and employ -B option.

**Comment by dannagle** [comment]
The interface is not chosen by Packet Sender, it is chosen by the OS. Packet Sender tells the OS it wants a socket that can route to 255.255.255.255, and the OS says "here you go." 

Having it be "all" would underneath be a socket for every IP (effectively the loop you just described). I agree that could be useful, but I feel it would be too error prone. The target interfaces could be virtual adapters (VMware, Docker, WSL, etc) that are ultimately routed to a single network. Packet Sender could very easily get confused and send out 5 packets when only 1 was desired because it got confused when probing the setup. 

The -B option is the intended solution to help this.

**Comment by lazna** [comment]
Understand. 

Just FYI: You could easily loop through physical adapters only, in batch file this is much more complicated.

**Comment by lazna** [comment]
One more note. Sending broadcast packet to specific port and waiting for undiscovered devices response with its IP address. Just found if interface have more than one IP assigned, and I am choosing for -B parameter random IP from list, not all devices respond for some reason. So I need to loop through all physical adapters, than loop through all its IPs and start PS for each IP again.

If PS have option for such behavior, it will be really nice.
